File: android/locxx-rules/src/main/kotlin/com/locxx/rules/LocXXModels.kt
package com.locxx.rules

enum class RowId {
    RED,
    YELLOW,
    GREEN,
    BLUE
}

enum class DieColor {
    WHITE1,
    WHITE2,
    RED,
    YELLOW,
    GREEN,
    BLUE
}

fun RowId.dieColor(): DieColor = when (this) {
    RowId.RED -> DieColor.RED
    RowId.YELLOW -> DieColor.YELLOW
    RowId.GREEN -> DieColor.GREEN
    RowId.BLUE -> DieColor.BLUE
}

/** Values left-to-right on the score sheet for each row (Qwixx layout). */
fun rowValues(row: RowId): IntArray = when (row) {
    RowId.RED, RowId.YELLOW -> IntArray(11) { i -> i + 2 }
    RowId.GREEN, RowId.BLUE -> IntArray(11) { i -> 12 - i }
}

data class PlayerRowState(
    /** Indices along [rowValues] that have been crossed (paper-accurate; gaps = skipped). */
    val crossedIndices: Set<Int> = emptySet(),
    val locked: Boolean = false
) {
    val crossCount: Int get() = crossedIndices.size
    val maxCrossedIndex: Int get() = crossedIndices.maxOrNull() ?: -1
}

data class PlayerSheet(
    val rows: Map<RowId, PlayerRowState> = enumValues<RowId>().associateWith { PlayerRowState() },
    val penalties: Int = 0
)

data class DiceRoll(
    val white1: Int,
    val white2: Int,
    val red: Int,
    val yellow: Int,
    val green: Int,
    val blue: Int
) {
    fun whiteSum(): Int = white1 + white2

    fun value(d: DieColor): Int = when (d) {
        DieColor.WHITE1 -> white1
        DieColor.WHITE2 -> white2
        DieColor.RED -> red
        DieColor.YELLOW -> yellow
        DieColor.GREEN -> green
        DieColor.BLUE -> blue
    }
}

File: ios/LocXXRules/Sources/LocXXRules/LocXXModels.swift
import Foundation

public enum RowId: CaseIterable, Hashable, Sendable {
    case red, yellow, green, blue
}

public enum DieColor: CaseIterable, Hashable, Sendable {
    case white1, white2, red, yellow, green, blue
}

public extension RowId {
    func dieColor() -> DieColor {
        switch self {
        case .red: return .red
        case .yellow: return .yellow
        case .green: return .green
        case .blue: return .blue
        }
    }
}

/// Values left-to-right on the score sheet for each row (Qwixx layout).
public func rowValues(_ row: RowId) -> [Int] {
    switch row {
    case .red, .yellow:
        return Array(2 ... 12)
    case .green, .blue:
        return Array((2 ... 12).reversed())
    }
}

public struct PlayerRowState: Equatable, Sendable {
    public var lastCrossedIndex: Int
    public var crossCount: Int
    public var locked: Bool

    public init(lastCrossedIndex: Int = -1, crossCount: Int = 0, locked: Bool = false) {
        self.lastCrossedIndex = lastCrossedIndex
        self.crossCount = crossCount
        self.locked = locked
    }
}

public struct PlayerSheet: Equatable, Sendable {
    public var rows: [RowId: PlayerRowState]
    public var penalties: Int

    public init(rows: [RowId: PlayerRowState] = Dictionary(uniqueKeysWithValues: RowId.allCases.map { ($0, PlayerRowState()) }), penalties: Int = 0) {
        self.rows = rows
        self.penalties = penalties
    }
}

public struct DiceRoll: Equatable, Sendable {
    public let white1: Int
    public let white2: Int
    public let red: Int
    public let yellow: Int
    public let green: Int
    public let blue: Int

    public init(white1: Int, white2: Int, red: Int, yellow: Int, green: Int, blue: Int) {
        self.white1 = white1
        self.white2 = white2
        self.red = red
        self.yellow = yellow
        self.green = green
        self.blue = blue
    }

    public func whiteSum() -> Int { white1 + white2 }

    public func value(_ d: DieColor) -> Int {
        switch d {
        case .white1: return white1
        case .white2: return white2
        case .red: return red
        case .yellow: return yellow
        case .green: return green
        case .blue: return blue
        }
    }
}
